File: javascript/data-structures/linked-list.js
// === Linked List: Functional ===
function link(head, tail = null) {
    const list = Object.create(null);
    list["head"] = head;
    list["tail"] = tail;
    return list;
}
function traverse(list, action) {
    if (list.tail === null) {
        return link(action(list.head));
    }
    return link(action(list.head), traverse(list.tail, action));
}
// === Linked List Construction ===
const list = link(3, link(2, link(1)));
const square_list = traverse(list, (x) => x * x);

// === list ===
// {
//     head: 3,
//     tail: {
//         head: 2,
//         tail: {
//             head: 1,
//             tail: null,
//         },
//     },
// };

// === square_list ===
// {
//     head: 9,
//     tail: {
//         head: 4,
//         tail: {
//             head: 1,
//             tail: null,
//         },
//     },
// };

// === Linked List: Object Oriented ===
// Use prototypal inheritance. Every link in the list need not have its own methods.
const proto_link = {
    traverse: function (action) {
        this.head = action(this.head);
        if (this.tail !== null) {
            this.tail.traverse(action);
        }
    },
};

// === Linked List ===
function prolink(head, tail = null) {
    const list = Object.create(proto_link);
    list["head"] = head;
    list["tail"] = tail;
    return list;
}

// === Linked List Construction ===
const prolist = prolink(3, prolink(2, prolink(1)));
prolist.traverse((x) => x * x);
// === prolist: post-traversal ===
// {
//     head: 9,
//     tail: {
//         head: 4,
//         tail: {
//             head: 1,
//             tail: null,
//         },
//     },
// };

File: python/data-structures/linked-list.py
# === Linked List: Object Oriented ===
class Link:
    def __init__(self, head, tail = None):
        self.head = head
        self.tail = tail

    def traverse(self, action):
        self.head = action(self.head)
        if self.tail != None:
            self.tail.traverse(action)

# === Examples ====
linked_list = Link(1, Link(2, Link(3)))
linked_list.traverse(lambda x: x * x)

# === Linked List: Functional ===
def link(head, tail = None):
    linker = dict()
    linker["head"] = head
    linker["tail"] = tail
    return linker

def traverse(linked_list, action):
    if linked_list["tail"] == None:
        return link(action(linked_list["head"]))
    else:
        return link(action(linked_list["head"]), traverse(linked_list["tail"], action))

# === Examples ===
fun_list = link(1, link(2, link(3)))
square_list = traverse(fun_list, lambda x: x * x)
